datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1322,6 +1323,23 @@ private void performPartitionMovement(Long notifyTimestamp) {
     }
   }
 
+  private void validateNewAssignment(Map<String, List<DatastreamTask>> newAssignmentsByInstance) {
+    if (_config.getMaxDatastreamTasksPerInstance() > 0) {
+      // If the cluster was configured to limit the max tasks per instance, check if any instances have a higher
+      // number of tasks than expected, and fail the leader assignment on violation of this limit. This can be useful
+      // to prevent other issues such as OOMs due to high memory usage which may be seen if we exceed the supportable
+      // number of tasks per instance.
+      Map<String, Integer> instancesWithTaskCountAboveThreshold = newAssignmentsByInstance.entrySet().stream()
```

Comment:
Should we also consider doing assignments only if a "quorum" number of instances are participating in the membership?

Reply:
Discussed this with @DEEPTHIKORAT offline. One problem with using a "quorum" is that it only helps catch cases when the number of tasks within the cluster are within the limit overall, but for some reason some of the instances are down.

There can be another case, where someone adds a new datastream or bumps the maxTasks in such a way that some of instances get a higher number of tasks than the instance can support without running into problems (e.g. OOMs). In such cases, the right thing to do is to add additional instances to the cluster such that the higher capacity can be supported.

So for now using a limit on tasks per instance is sufficient.